apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/ExponentPackageLogger.kt

```diff
@@ -0,0 +1,131 @@
+    .writeTimeout(10, TimeUnit.SECONDS)
+    .readTimeout(0, TimeUnit.MINUTES) // Disable timeouts for read
+    .build()
+
+  private var webSocket: WebSocket? = null
+  private var connected: Boolean = false
+  private var pendingMessage: String? = null
+  private var logLevel: LogLevel = LogLevel.INFO
+
+  private suspend fun sendMessage(message: String, logLevel: LogLevel) {
+    pendingMessage = message
+    this.logLevel = logLevel
+
+    val request: Request = Request.Builder().url(getMessageSocketUrl(appUrl)).build()
+
+    val latch = CompletableDeferred<Unit>()
+
+    okHttpClient.newWebSocket(
+      request,
+      object : WebSocketListener() {
+        override fun onOpen(webSocket: WebSocket, response: Response) {
+          connected = true
+          this@ExponentPackageLogger.webSocket = webSocket
+          sendPendingMessage()
+          latch.complete(Unit)
+        }
+
+        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+          Log.e(TAG, "Error opening web socket to url: ${getMessageSocketUrl(appUrl)}")
+          latch.completeExceptionally(t)
+        }
+
+        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+          connected = false
+        }
+      }
+    )
+
+    latch.await()
+  }
+
+  private fun sendPendingMessage() {
+    if (!connected) {
+      return
+    }
+
+    if (pendingMessage != null && webSocket != null) {
+      val msg = mapOf(
+        "type" to "log",
+        "level" to when (logLevel) {
+          LogLevel.WARNING -> "warn"
+          LogLevel.ERROR -> "error"
+          LogLevel.INFO -> "info"
+        },
+        "data" to arrayOf(pendingMessage)
+      )
+
+      try {
+        val json = JSONObject(msg).toString()
+        webSocket?.send(json)
+      } catch (e: JSONException) {
+        Log.e(TAG, "Failed to stringify message: ${e.message}")
+      }
+      pendingMessage = null
+      try {
+        webSocket?.close(1000, "End of session")
+      } catch (e: Exception) {
+        // swallow, no need to handle it here
+      }
+      webSocket = null
+      connected = false
+    }
+  }
+
+  private fun getMessageSocketUrl(appUrl: Uri): String {
+    val host = appUrl.host ?: "localhost"
+    val port = if (appUrl.port > 0) appUrl.port else 8081
+    val scheme = if (appUrl.scheme == "https") "wss" else "ws"
+    return "$scheme://$host:$port/hot"
+  }
+
+  companion object {
+    private val TAG = ExponentPackageLogger::class.java.simpleName
+
+    /**
+     * Sends a message to the packager with the given url.
+     * @param appUrl packager url
+     * @param message message to send
+     * @param logLevel log level to use, will be sent to the packager which can format or
+     * display different log levels in different ways.
+     */
+    fun send(appUrl: String, message: String, logLevel: LogLevel) {
+      val logger = ExponentPackageLogger(appUrl.toUri())
+      CoroutineScope(Dispatchers.IO).launch {
```

Comment:
lean toward to turn this function as `suspend fun send(...`) and remove the `CoroutineScope(Dispatchers.IO).launch {` here. let the callsite to launch the coroutine and it could be better to cancel this coroutine. for example, we can use the exisiting `kernelScope` coroutine scope + `withContext(Dispatchers.IO)`

Reply:
Good point!